### ISSUE TITLE
Cannot find module "wavesurfer.js" when using SSR

### ISSUE THREAD

**Issue by vinzentt** [mspae/react-wavesurfer#21]
Hi,

Thanks for this project.

I'm trying to use this component after 'npm install react-wavesurfer' it says 
Cannot find module "wavesurfer.js"
and after a npm i wavesurfer.js i'm getting a window is not defined...

Any Idea on how to make it work?

Thx.

**Comment by mspae** [comment]
Hello @vinzentt – Do you use a bundling system like webpack?

Can you post the exact error (i.e. where the window is not defined error appears)?

wavesurfer.js needs to be installed in the project and exposed on the global object.

With webpack this means you need to add this loader in the webpack.config.js. This should be documented better, my bad.

```javascript
{
  test: require.resolve("wavesurfer.js"),
  loader: "expose?WaveSurfer"
}
```

**Comment by mspae** [comment]
I have updated the documentation to include a mention of the specifics of this library. https://github.com/mspae/react-wavesurfer/commit/1fc610648e5e92fdda34bfcacb14e3ff47f48f74

The list is by no means complete or as intelligible as I would like, so if you or anyone else has any additional experience or can formulate it in a better way please go ahead :yum:

**Comment by vinzentt** [comment]
Hi @mspae 

thanks for your answer. First i must tell you I'm completely new to react and i'm using a [starter kit](https://github.com/Hashnode/mern-starter) with MERN architecture and yes it uses webpack. 

So I added the lines you told me in webpack.config.js but I still have the window is not defined issue. As far as I understand, it's because the component is rendered on server side and it shouldn't. So will try to change the way I include my components.

Thanks again.

**Comment by n-sakib** [comment]
Hi, can you share a mroe detailed webpack-config? I'm also pretty new and I also can't understand this.

**Comment by mspae** [comment]
@vinzentt Hi, so the wavesurfer library needs a browser to be able to render a canvas with a wave etc. I think a solution for your problem would be to not render the Wavesurfer component when rendering on the server and do it when the client is rendering the page. I have no experience with SSR, but I assume there is some flag you can use to check in which environment the code is currently being executed.

**Comment by mspae** [comment]
– Note that this is unrelated to the issue in this thread, which is that react-wavesurfer needs to be aware if it is being used in a server-side rendering context –

@n-sakib – the webpack config I posted in the comment above is only necessary if you intend to use a react wavesurfer plugin. For "normal" usage (without any plugin components, only the basic player) you just have to install `react-wavesurfer` and `wavesurfer.js` (and save them as (dev-)dependencies) and then you can use them like this:

```javascript
import React, { Component } from 'react';
import Wavesurfer from 'react-wavesurfer';

export default class App extends Component {
  render() {
    return (
      <div>
        <Wavesurfer audioFile="https://ia902606.us.archive.org/35/items/shortpoetry_047_librivox/song_cjrg_teasdale_64kb.mp3" />
      </div>
    );
  }
}
```

If you want to use a plugin component, you need to add this custom webpack configuration:
```javascript
module: {
    loaders: [
      // … other loaders go here
      {
        test: require.resolve("wavesurfer.js"),
        loader: "expose?WaveSurfer"
      }
    ]
}
```

Then, in when you use the component, simply do:
```javascript
import React, { Component } from 'react';
import Wavesurfer from 'react-wavesurfer';
import Timeline from 'react-wavesurfer/lib/plugins/timeline.js';

export default class App extends Component {
  render() {
    return (
      <div>
        <Wavesurfer audioFile="https://ia902606.us.archive.org/35/items/shortpoetry_047_librivox/song_cjrg_teasdale_64kb.mp3">
          <Timeline />
        </Wavesurfer>
      </div>
    );
  }
}
```

**Comment by mxmtsk** [comment]
I stumbled across the same error. I'm using it with [react-starter-kit](https://github.com/kriasoft/react-starter-kit) ☹️

**Comment by mspae** [comment]
I have researched the issue. The issue is that `wavesurfer.js` expects the window to be defined before it is being called. (`defaultParams.pixelRatio` is calculated by using `window.devicePixelRatio` among others).

There is unfortunately no quick fix I can add to `react-wavesurfer` since this code is being run before the component is even being instantiated. I will open a PR on `wavesurfer.js` to adress this issue. In the mean time you could try having different builds for server and client. (remove `wavesurfer.js` on the server build and when you call the component check that window is defined)

**Comment by mxmtsk** [comment]
@mspae https://github.com/katspaugh/wavesurfer.js/issues/751 looks like it would render initially when you set the pixel ratio to 1. I couldn't test, because I'm getting a big error in my webpack configuration when I try to import wavesurfer (have to figure out why) but maybe you or @vinzentt could try out

**Comment by mspae** [comment]
https://github.com/katspaugh/wavesurfer.js/issues/751#issuecomment-227207301

> Then it's an easy fix. Set the default pixel ratio to 1 and in the init function conditionally assign to window.devicePixelRatio if no param was passed.

So I think what this would look like:

```javascript
if (!window) {
  // mock window object if we're doing SSR
  window = {
    devicePixelRatio: 1
  };
}
```

It might work, but it's still a bit hackish. Let me know if it works!

**Comment by vinzentt** [comment]
Hi all,

sorry for my late answer. I ended up with this solution: I'm requiring (ie importing) Wavesurfer only in the render function of my parent component. So I did sthg like this:

```
const WavesurferComponent = require('../../components/Wavesurfer').default;
wavesurferElement = 
        <WavesurferComponent ref="wavesurfer" audioUrl={this.state.audioUrl} />
```

And in my wavesurfer component I'm simply importing react-wavesurfer:
`import Wavesurfer from 'react-wavesurfer';`

This way, both my component and react-wavesurfer are imported on client side only, so no more problem with the undefined window. Tell me if you found a better solution.

Thanks for this component!
